content/blog/trungnt2910/2024-08-11_[gsoc_2024]_debugging_progress_3.md

```diff
@@ -0,0 +1,370 @@
+
+This version, however, still occasionally cause hangs in GDB. This happens for some asynchronous
+events like `THREAD_EXITED`. The main thread might read this event right after consuming the
+previous one. At the same time, the worker thread sees the port as readable, removes it from the
+list, and marks the `pipe_to_event_loop`. The main thread would wake up again, only to find out that
+there actually are no new events and not re-adding the discarded port. This makes the current team's
+port "lost" to the worker thread.
+
+#### Third Version
+
+This is the current version of the worker thread logic.
+
+```cpp
+// Worker thread loop
+wait_for_ports = true;
+while (true)
+{
+  build a list of ports
+
+  if (wait_for_ports)
+    wait_for_objects(pipe_to_worker + the ports);
+  else
+    wait_for_objects(pipe_to_worker)
+
+  if (pipe_to_worker is marked)
+  {
+    flush pipe to worker
+    wait_for_ports = true;
+  }
+  else
+    wait_for_ports = false;
+
+  remove invalid ports
+
+  if (any ports are readable)
+    pipe_to_event_loop.mark();
+}
+```
+
+To solve the "lost port" issue, we stop removing valid ports altogether. To prevent the initial
+issue where the worker thread runs into an infinite loop, we now only allow the thread to wait for
+ports when signalled by the main thread.
+
+The `pipe_to_worker` is now used for general signalling that the worker thread needs to either
+refresh its port list or start waiting for ports again. It is marked on `resume` and `wait`, similar
+to the second version.
+
+Unlike the second version, the this version no longer re-adds ports in `resume` and `wait`, reducing
+mutex usage in the event thread.
+
+This version has been tested with 100 runs of `Debugger`. So far I have not encountered any further
+issue.
+
+### Executable Relocation
+
+In the previous blog, I have mentioned using `disable_breakpoints_before_startup` to mute
+breakpoint-related functions before the main executable gets properly relocated.
+
+However, it turns out that
+[some components](https://github.com/trungnt2910/gdb-haiku/issues/4#issuecomment-2262975347) in GDB
+does not respect the startup flag. Therefore, executable relocation has now been pushed as early as
+possible, right after startup for inferiors we `fork`ed ourselves.
+
+### Memory Access
+
+#### Target Memory Regions
+
+As previously mentioned, the regions reported by `info mem` determines whether GDB would attempt
+to read/write a certain address. They have to be manually refreshed by the native facility.
+
+Previously, these regions are only invalidated when an image event or `exec` occurs. However, there
+are various more reasons mappings can change: threads being created, `[*]_area` or `mmap` called,
+and so on. Without an updated view,
+[stack traces and variable inspection](https://github.com/trungnt2910/gdb-haiku/issues/2) will not
+work correctly.
+
+The solution I have chosen is to invalidate everything whenever an event is read in `wait`. This
```

Comment:
Hmm. Maybe this is part of the slowness of attaching to WebKit? It does use hundreds of areas after all. (Did you try attaching to the existing single-process WebKit and see how long that takes?)

Reply:
Yesterday it was broken. Just saw the solution on the forum, will try that later.

Reply:
Can't wait long enough to see how long, but attaching another `gdb` to the running `gdb` reveals issues with C++ name demangling and multithreaded dynamic memory allocation.